File: Excel/603 Challenge.py
import pandas as pd

path = "603 Missing Numbers.xlsx"
input = pd.read_excel(path, usecols="A:C", skiprows=1, nrows=4, header=None)
test = pd.read_excel(path, usecols="E", nrows=6).astype("str")

V1 = input.values.flatten()
missing_nums = sorted(set(range(min(V1), max(V1) + 1)) - set(V1))

df_missing = pd.DataFrame(missing_nums, columns=["Missing Numbers"])
df_missing['index'] = (df_missing.diff() != 1).cumsum()

df_missing = df_missing.groupby('index')['Missing Numbers'].apply(
    lambda group: str(group.iloc[0]) if len(group) == 1 else f"{group.iloc[0]}-{group.iloc[-1]}"
).reset_index(drop=True)

print(df_missing.equals(test['Answer Expected'])) # True

File: Excel/600-699/603/603 Challenge.R
library(tidyverse)
library(readxl)

path = "Excel/603 Missing Numbers.xlsx"
input = read_excel(path, range = "A2:C5", col_names = F) %>% as.matrix()
test  = read_excel(path, range = "E1:E6")

V1 = input %>%
  as.vector() 
V2 = min(V1):max(V1)

result = data.frame(nums = setdiff(V2, V1)) %>%
  mutate(group = cumsum(c(1, diff(nums) != 1))) %>%
  summarise(nums = ifelse(n() > 1, 
                          paste0(min(nums), "-", max(nums)), 
                          as.character(nums)), .by = group) 

all.equal(result$nums, test$`Answer Expected`)
#> [1] TRUE